src/ucp/core/ucp_rkey.h

```diff
@@ -84,28 +96,39 @@ typedef struct {
  * The array itself contains only the MDs specified in md_map, without gaps.
  */
 typedef struct ucp_rkey {
-    /* cached values for the most recent endpoint configuration */
-    struct {
-        ucp_worker_cfg_index_t    ep_cfg_index; /* EP configuration relevant for the cache */
-        ucp_lane_index_t          rma_lane;     /* Lane to use for RMAs */
-        ucp_lane_index_t          amo_lane;     /* Lane to use for AMOs */
-        ssize_t                   max_put_short;/* Cached value of max_put_short */
-        uct_rkey_t                rma_rkey;     /* Key to use for RMAs */
-        uct_rkey_t                amo_rkey;     /* Key to use for AMOs */
-        ucp_amo_proto_t           *amo_proto;   /* Protocol for AMOs */
-        ucp_rma_proto_t           *rma_proto;   /* Protocol for RMAs */
-    } cache;
-    ucp_md_map_t                  md_map;       /* Which *remote* MDs have valid memory handles */
-    ucs_memory_type_t             mem_type;     /* Memory type of remote key memory */
-    uint8_t                       flags;        /* Rkey flags */
-    ucp_worker_cfg_index_t        cfg_index;    /* Rkey configuration index */
+    union {
+        /* cached values for the most recent endpoint configuration */
+        struct {
+            uint8_t                   flags;          /* Rkey flags */
+            ucp_worker_cfg_index_t    ep_cfg_index;   /* EP configuration relevant for the cache */
+            ucp_lane_index_t          rma_lane;       /* Lane to use for RMAs */
+            ucp_lane_index_t          amo_lane;       /* Lane to use for AMOs */
+            ucp_rkey_proto_index_t    amo_proto_index;/* Protocol for AMOs */
+            ucp_rkey_proto_index_t    rma_proto_index;/* Protocol for RMAs */
+            uct_rkey_t                rma_rkey;       /* Key to use for RMAs */
+            uct_rkey_t                amo_rkey;       /* Key to use for AMOs */
+            ssize_t                   max_put_short;  /* Cached value of max_put_short */
```

Comment:
maybe reduce it to int16 or int32?

Reply:
did you get 40 bytes without removing it? Maybe removing it at all is worth considering?

Reply:
40 bytes without removing.
With the `rkey_t` `max_put_short` uses only here: https://github.com/openucx/ucx/blob/master/src/ucp/rma/rma_send.c#L282

Seems like `max_put_short`  cannot be reduced to int16 or 32, it cached the following value:
```C
typedef struct ucp_ep_rma_config {
    ssize_t                max_put_short;    /* Maximal payload of put short */
```
@yosefe what do you think if it is removed from `rkey_t`?